cmd/bm-client/handlers/otp.go

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2020 BitMaelum Authors
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy of
+// this software and associated documentation files (the "Software"), to deal in
+// the Software without restriction, including without limitation the rights to
+// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software is furnished to do so,
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+package handlers
+
+import (
+	"crypto/aes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/bitmaelum/bitmaelum-suite/internal"
+	"github.com/bitmaelum/bitmaelum-suite/internal/container"
+	"github.com/bitmaelum/bitmaelum-suite/pkg/bmcrypto"
+	"github.com/bitmaelum/bitmaelum-suite/pkg/hash"
+	"github.com/olekukonko/tablewriter"
+	"github.com/sirupsen/logrus"
+)
+
+// OtpGenerate will generate an OTP valid for otpServer
+func OtpGenerate(info *internal.AccountInfo, otpServer *string) {
+	// Get
+	recs, err := net.LookupTXT("_bitmaelum." + *otpServer)
+	if err != nil {
+
+		logrus.Fatal(err)
+	}
+
+	resolver := container.GetResolveService()
+
+	for _, txt := range recs {
+		orgHash, err := hash.NewFromHash(strings.ToLower(txt))
+		if err != nil {
+			continue
+		}
+
+		oi, err := resolver.ResolveOrganisation(*orgHash)
+		if err == nil {
+			// Generate secret on the client and compute OTP
+			secret, err := bmcrypto.KeyExchange(info.PrivKey, oi.PublicKey)
+			if err != nil {
+				logrus.Fatal(err)
+			}
+
+			printOtpLoop(secret, *otpServer)
+
+			return
+		}
+	}
+
+	logrus.Fatal(errors.New("public key not found for " + *otpServer))
+}
+
+func printOtpLoop(secret []byte, server string) {
+
+	for {
+		otp := computeOTPFromSecret(secret, 8)
+
+		v := 60 - ((time.Now().UnixNano() - getLastMinuteTimestamp()) / int64(time.Second))
+
+		table := tablewriter.NewWriter(os.Stdout)
+		table.SetHeader([]string{"OTP", "Server", "Valid for"})
+		table.Append([]string{otp, server, fmt.Sprintf("%d", v)})
+
+		table.Render()
+
+		<-time.After(1 * time.Second)
+	}
+}
+
+func getLastMinuteTimestamp() int64 {
+	t := time.Now()
+	rounded := time.Date(t.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), 0, 0, time.UTC)
+	return rounded.UnixNano()
+}
+
+func computeOTPFromSecret(secret []byte, otpLength int) string {
```

Comment:
I really think this is too complex. We have an exchanged (long enough) secret, with which we can use a standard hmac with. From that hmac we could do a modulus on the wanted length `(% 10^len)`.  I think this makes things a bit easier. Also, i would like to move to a 30 second block instead of a minute  (you can use `floor(time/30) * 30` to quickly truncate to a given time block

Reply:
You're right, I did implement both of your suggestions